lib/libzpool/uzfs_io.c

```diff
@@ -513,3 +513,81 @@ uzfs_update_metadata_granularity(zvol_state_t *zv, uint64_t tgt_block_size)
 	zv->zv_metavolblocksize = tgt_block_size;
 	return (0);
 }
+
+int
+uzfs_unmap_data(zvol_state_t *zv, uint64_t offset, uint64_t len,
+    blk_metadata_t *metadata, boolean_t is_rebuild)
+{
+	int error = 0;
+	uint32_t count = 0;
+	list_t *chunk_io = NULL;
+	uint64_t orig_offset = offset;
+	uint64_t end = len + offset;
+	uint64_t wrote = 0;
+	uint64_t blocksize = zv->zv_volblocksize;
+	rl_t *rl;
+	uint64_t len_in_first_aligned_block = 0;
+
+	/*
+	 * If trying IO on fresh zvol before metadata granularity is set return
+	 * error.
+	 */
+
+	if (zv->zv_metavolblocksize == 0)
+		return (EINVAL);
+
+	ASSERT(metadata);
+	ASSERT3P(zv->zv_metavolblocksize, !=, 0);
+	if (!IS_P2ALIGNED(offset, zv->zv_metavolblocksize) ||
+	    !IS_P2ALIGNED(len, zv->zv_metavolblocksize) ||
+	    len == 0)
+		return (EINVAL);
+
+	if (offset + len > zv->zv_volsize || offset > zv->zv_volsize)
+		return (EINVAL);
+
+	rl = zfs_range_lock(&zv->zv_range_lock, offset, len, RL_WRITER);
+
+	if (!is_rebuild) {
+		metadata->io_num = CREATE_UNMAP_IOSEQ(metadata->io_num);
+	} else {
+		if (!CHECK_UNMAP_IOSEQ(metadata->io_num))
+			return (EINVAL);
+
+		VERIFY(ZVOL_IS_DEGRADED(zv) && (ZVOL_IS_REBUILDING(zv) ||
+		    ZVOL_IS_REBUILDING_ERRORED(zv)));
+		count = uzfs_get_nonoverlapping_ondisk_blks(zv, offset,
+		    len, metadata, (void **)&chunk_io);
+		if (!count)
+			goto out;
+chunk_io:
+			GET_NEXT_CHUNK(chunk_io, offset, len, end);
+			wrote = offset - orig_offset;
+			CHECK_FIRST_ALIGNED_BLOCK(
```

Comment:
alignment here is not required, as, dnode_free_range takes care of it

Reply:
ok